proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,527 @@
+are configured on the GitHub App or the Fine-grained PAT. To use the above APIs, the client or PAT must be configured
+to allow reading the organization's memberships. Additionally, the org owners may need to approve the GitHub App
+or Fine-grained PAT's permissions for that org.
+
+All REST API calls are subject to per-user
+[rate limits](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28).
+
+## Proposal
+
+### Goals and Non-goals
+
+Goals:
+
+- Add new GitHubIdentityProvider resource which allows Supervisor admins to configure GitHub
+  as an identity provider. Offer configuration options for extracting usernames and group memberships, as well
+  as for preventing authentication unless a user belongs to a certain GitHub organization.
+- Allow GitHubIdentityProvider(s) to be configured by admins on FederationDomains, and enable all features of
+  FederationDomains on them (e.g. identity transformations).
+- Allow end users to use a web browser-based authentication flow, similar to other supported identity providers,
+  where the end user only needs to authenticate using a web browser once a day.
+- Allow end users to use a CLI-based authentication flow which does not require a web browser,
+  similar to other supported identity providers. This enables using Pinniped authentication for scripting
+  and CI purposes. Allow the admin to disable this feature if they prefer.
+- Regardless of which flow the user uses to authenticate, regularly check with GitHub to see if the user should
+  be allowed to continue their session. If the user's GitHub tokens were revoked, or if the user no longer belongs
+  to the configured GitHub organization(s), then end their Pinniped Supervisor session to block their continued
+  access to Kubernetes clusters.
+- Regardless of which flow the user uses to authenticate, regularly check with GitHub to see if the user's group
+  memberships have changed and update them accordingly.
+- Any client of the Supervisor should be able to authenticate GitHub users using the web browser-based
+  authentication flow, including the Pinniped CLI and third-party clients registered with the Supervisor as OIDCClients.
+- The IDP chooser page and the IDP discovery endpoints should both include the configured GitHub identity provider(s)
+  for the FederationDomain in the lists that they return.
+- Avoid requiring any changes to the Pinniped CLI or other clients. From a client's point of view,
+  which external identity provider is used for authentication should be transparent.
+
+### Specification / How it Solves the Use Cases
+
+#### New GitHubIdentityProvider resource
+
+Add a new CRD called GitHubIdentityProvider. Here is an example:
+
+```yaml
+apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+kind: GitHubIdentityProvider
+metadata:
+  namespace: supervisor
+  name: github
+spec:
+  github_api:
```

Comment:
This should be `gitHubAPI`.

Reply:
Actually, names TBD. See https://github.com/vmware-tanzu/pinniped/pull/1900#discussion_r1539473295

TODO: update this name when decided